### PR TITLE
Enable USB Full speed Device controller on LPC55S36 and LPC55S59

### PR DESCRIPTION
**Pull request by mmahadevan108**
Add support for the Full Speed device controller on LPC55S36 and LPC55S59

### CODE REVIEW

File: drivers/usb/device/Kconfig [hunk 1]
```diff
@@ -158,7 +158,7 @@ config USB_DC_NXP_EHCI
 
 config USB_DC_NXP_LPCIP3511
 	bool "LPC USB Device Controller"
-	select USB_DC_HAS_HS_SUPPORT
+	select USB_DC_HAS_HS_SUPPORT if "$(dt_nodelabel_enabled,usbhs)"
```
Comment (decsny) [inline drivers/usb/device/Kconfig:161]: is there a way to do this without the need for a specifically labelled node in DT?

File: soc/arm/nxp_lpc/lpc55xxx/Kconfig.soc [hunk 2]
```diff
@@ -151,4 +151,12 @@ config LPC55XXX_SRAM_CLOCKS
 	  By default, CMSIS SystemInit will enable the clock to these RAM banks.
 	  Disable this Kconfig to leave the ram banks untouched out of reset.
 
+config LPC55XXX_USB_RAM
+	bool "Is USB RAM Present?"
```
Comment (decsny) [inline soc/arm/nxp_lpc/lpc55xxx/Kconfig.soc:155]: I think this kconfig should not have a menu prompt. Also, would this be more appropriate as a devicetree property?

File: boards/arm/lpcxpresso55s36/pinmux.c [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+static int lpcxpresso55s36_pinmux_init(void)
+{
+	int err;
+
+	/* Declare pin configuration state for USB pin here */
+	PINCTRL_DT_DEFINE(DT_NODELABEL(usbfs));
+
+	/* Apply pinctrl state directly, since there is no flexio device driver */
+	err = pinctrl_apply_state(PINCTRL_DT_DEV_CONFIG_GET(DT_NODELABEL(usbfs)),
```
Comment (decsny) [inline boards/arm/lpcxpresso55s36/pinmux.c:19]: why does this not work from DT? there needs to be a flexio driver?

File: dts/arm/nxp/nxp_lpc55S6x_common.dtsi [hunk 4]
```diff
@@ -314,7 +314,17 @@
 		#io-channel-cells = <1>;
 	};
 
-	usbhs: usbhs@144000 {
+	usbfs: usbfs@84000 {
+		compatible = "nxp,mcux-usbd";
```
Comment (decsny) [inline dts/arm/nxp/nxp_lpc55S6x_common.dtsi:318]: in the commit message here is says "also fix the usb ram mapping" but that seems to happen in a later commit

File: soc/arm/nxp_lpc/lpc55xxx/soc.c [hunk 5]
```diff
@@ -155,6 +155,41 @@ static ALWAYS_INLINE void clock_init(void)
...
+	 * usb host register
+	 */
+	*((uint32_t *)(USBFSH_BASE + 0x5C)) |= USBFSH_PORTMODE_DEV_ENABLE_MASK;
+	/* disable usb0 host clock */
+	CLOCK_DisableClock(kCLOCK_Usbhsl0);
+
+	/* enable USB IP clock */
+	CLOCK_EnableUsbfs0DeviceClock(kCLOCK_UsbfsSrcFro, CLOCK_GetFroHfFreq());
+#if defined(FSL_FEATURE_USB_USB_RAM) && (FSL_FEATURE_USB_USB_RAM)
+	for (int i = 0; i < FSL_FEATURE_USB_USB_RAM; i++) {
```
Comment (danieldegrasse) [inline soc/arm/nxp_lpc/lpc55xxx/soc.c:185]: Why not use `memset` here?

File: boards/arm/lpcxpresso55s36/pinmux.c (hunk 3 above)
Comment (DerekSnell) [reply boards/arm/lpcxpresso55s36/pinmux.c:19]: FlexIO is not related to this, so I think that is an error in the comment

File: boards/arm/lpcxpresso55s36/pinmux.c [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+#include <zephyr/init.h>
+#include <zephyr/drivers/pinctrl.h>
+#include "fsl_common.h"
+
+static int lpcxpresso55s36_pinmux_init(void)
+{
+	int err;
+
+	/* Declare pin configuration state for USB pin here */
+	PINCTRL_DT_DEFINE(DT_NODELABEL(usbfs));
```
Comment (danieldegrasse) [inline boards/arm/lpcxpresso55s36/pinmux.c:16]: Is there any reason we could not declare the pin configuration state within the USB DC driver? Even though the driver does not use the device model, it should still be possible to locate this code there and call `pinctrl_apply_state` during init. If no pin control states are specified, the function will of course return an error, but other drivers like the [memc_mcux_flexspi](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/memc/memc_mcux_flexspi.c#L187) driver simply ignore the return code if it indicates the error was that no pin control state was present.

**Comment by imi415** [comment]
Hi @mmahadevan108, I have tested this PR on LPCXpresso55S36, it seems the code path with no dedicated USB SRAM has issue which will truncate CONTROL IN data if CONFIGURATION descriptor is above 64 bytes.

I believe this issue should not be related to this PR itself, however AFAIK LPC55(S)36 is the only device with USBFS support without USB SRAM. I believe submitting a issue for this should be more suitable, however since no existing devices will be impacted until this PR is merged, I have to put it here.

I use CDC ACM example, and the length of the CONFIGURATION is 75 bytes.

On a successful CONFIGURATION descriptor request, the transfer is looks like the following:
usb_success

When the USB SRAM is disabled, the descriptor is truncated and a ZLP is sent:
usb_fail

On LPC55S69, a simple way to reproduce this is disabling the macro `FSL_FEATURE_USBHSD_USB_RAM_BASE_ADDRESS`, which forces the driver to use allocated pool instead of dedicated SRAM, this way I can reproduce the same issue.

Failed USB enumeration log:
```
[ 5134.914456] usb 1-4.1.2: new full-speed USB device number 65 using xhci_hcd
[ 5135.532379] usb 1-4.1.2: config index 0 descriptor too short (expected 75, got 64)
[ 5135.532392] usb 1-4.1.2: config 1 has an invalid descriptor of length 7, skipping remainder of the config
[ 5135.532400] usb 1-4.1.2: config 1 interface 1 altsetting 0 has 0 endpoint descriptors, different from the interface descriptor's value: 2
[ 5135.916116] usb 1-4.1.2: New USB device found, idVendor=2fe3, idProduct=0001, bcdDevice= 3.03
[ 5135.916132] usb 1-4.1.2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[ 5135.916138] usb 1-4.1.2: Product: Zephyr CDC ACM sample
[ 5135.916142] usb 1-4.1.2: Manufacturer: ZEPHYR
[ 5135.916146] usb 1-4.1.2: SerialNumber: 0123456789ABCDEF
[ 5136.013519] cdc_acm: probe of 1-4.1.2:1.0 failed with error -22
```

File: boards/arm/lpcxpresso55s36/pinmux.c (hunk 6 above)
Comment (mmahadevan108) [reply boards/arm/lpcxpresso55s36/pinmux.c:16]: Thank you, I will add to the USB driver.

File: boards/arm/lpcxpresso55s36/pinmux.c (hunk 3 above)
Comment (mmahadevan108) [reply boards/arm/lpcxpresso55s36/pinmux.c:19]: This code has been deleted,

File: dts/arm/nxp/nxp_lpc55S6x_common.dtsi (hunk 4 above)
Comment (mmahadevan108) [reply dts/arm/nxp/nxp_lpc55S6x_common.dtsi:318]: Thanks for spotting this, I have updated the commit message.

File: soc/arm/nxp_lpc/lpc55xxx/soc.c (hunk 5 above)
Comment (mmahadevan108) [reply soc/arm/nxp_lpc/lpc55xxx/soc.c:185]: Will do.

File: soc/arm/nxp_lpc/lpc55xxx/Kconfig.soc (hunk 2 above)
Comment (mmahadevan108) [reply soc/arm/nxp_lpc/lpc55xxx/Kconfig.soc:155]: Removed the menu prompt. This is Kconfig as we need this to be available in the CMake file.

**Comment by mmahadevan108** [comment]
> Hi @mmahadevan108, I have tested this PR on LPCXpresso55S36, it seems the code path with no dedicated USB SRAM has issue which will truncate CONTROL IN data if CONFIGURATION descriptor is above 64 bytes.
> 
> I believe this issue should not be related to this PR itself, however AFAIK LPC55(S)36 is the only device with USBFS support without USB SRAM. I believe submitting a issue for this should be more suitable, however since no existing devices will be impacted until this PR is merged, I have to put it here.
> 
> I use CDC ACM example, and the length of the CONFIGURATION is 75 bytes.
> 
> On a successful CONFIGURATION descriptor request, the transfer is looks like the following: usb_success
> 
> When the USB SRAM is disabled, the descriptor is truncated and a ZLP is sent: usb_fail
> 
> On LPC55S69, a simple way to reproduce this is disabling the macro `FSL_FEATURE_USBHSD_USB_RAM_BASE_ADDRESS`, which forces the driver to use allocated pool instead of dedicated SRAM, this way I can reproduce the same issue.
> 
> Failed USB enumeration log:
> 
> ```
> [ 5134.914456] usb 1-4.1.2: new full-speed USB device number 65 using xhci_hcd
> [ 5135.532379] usb 1-4.1.2: config index 0 descriptor too short (expected 75, got 64)
> [ 5135.532392] usb 1-4.1.2: config 1 has an invalid descriptor of length 7, skipping remainder of the config
> [ 5135.532400] usb 1-4.1.2: config 1 interface 1 altsetting 0 has 0 endpoint descriptors, different from the interface descriptor's value: 2
> [ 5135.916116] usb 1-4.1.2: New USB device found, idVendor=2fe3, idProduct=0001, bcdDevice= 3.03
> [ 5135.916132] usb 1-4.1.2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
> [ 5135.916138] usb 1-4.1.2: Product: Zephyr CDC ACM sample
> [ 5135.916142] usb 1-4.1.2: Manufacturer: ZEPHYR
> [ 5135.916146] usb 1-4.1.2: SerialNumber: 0123456789ABCDEF
> [ 5136.013519] cdc_acm: probe of 1-4.1.2:1.0 failed with error -22
> ```

@imi415 Thank you for testing. I have updated the PR to fix this.

File: drivers/usb/device/Kconfig (hunk 1 above)
Comment (mmahadevan108) [reply drivers/usb/device/Kconfig:161]: I am not aware of a way.

**Comment by imi415** [comment]
> @imi415 Thank you for testing. I have updated the PR to fix this.
Hi @mmahadevan108, I have tested the new fixes, it works now. Thanks!

File: drivers/usb/device/usb_dc_mcux.c [hunk 7]
```diff
@@ -14,6 +14,7 @@
 #include <soc.h>
 #include <zephyr/device.h>
 #include <zephyr/kernel.h>
+#include <zephyr/drivers/pinctrl.h>
```
Comment (danieldegrasse) [inline drivers/usb/device/usb_dc_mcux.c:17]: Add `select PINCTRL` to the Kconfig that enables the USB driver, as this is how we insure pin control is enabled in other drivers: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/spi/Kconfig.mcux_lpspi#L11

File: dts/bindings/usb/nxp,mcux-usbd.yaml [hunk 8]
```diff
@@ -6,7 +6,7 @@ description: |
 
 compatible: "nxp,mcux-usbd"
 
-include: usb-ep.yaml
+include: [usb-ep.yaml, pinctrl-device.yaml]
```
Comment (danieldegrasse) [inline dts/bindings/usb/nxp,mcux-usbd.yaml:9]: This `pinctrl-device.yaml` should be updated in the commit enabling pinctrl for the USB driver

File: boards/arm/lpcxpresso55s69/pinmux.c [hunk 9]
```diff
@@ -11,7 +11,6 @@
 
 static int lpcxpresso_55s69_pinmux_init(void)
 {
-
```
Comment (danieldegrasse) [inline boards/arm/lpcxpresso55s69/pinmux.c:14]: Nit- this line deletion does not appear to be necessary

**Review by danieldegrasse — COMMENTED** [review]
A few small comments, but otherwise this PR looks good

**Review by decsny — APPROVED** [review]

**Review by danieldegrasse — APPROVED** [review]

**Review by jfischer-no — APPROVED** [review]

**Pull request merged**